From 1a4b20dccbbd5a9d01ce163720685eb3919aa001 Mon Sep 17 00:00:00 2001
From: Galaxy Book 12 Linux project <local@galaxybook12>
Date: Sat, 8 Aug 2026 08:05:00 +0200
Subject: [PATCH] platform/x86: int3472: map Galaxy Book 12 sensor rails

Samsung's Skylake firmware uses the original Intel Windows control-logic
names for its IMX258 power GPIOs: 0x0e is the autofocus rail, 0x0f is IO,
0x10 is Avdd and 0x11 is Core.  Linux otherwise ignores three of those GPIOs
and interprets 0x10 as DOVDD, leaving the sensor and actuator unpowered.

Map the three sensor functions to the supply names used by the mainline IMX258
driver and expose the AF line as an actuator-power GPIO. The mapping is
restricted to the SONY258A ACPI HID.
---
 drivers/platform/x86/intel/int3472/discrete.c | 27 +++++++++++++++++++++++++++
 1 file changed, 27 insertions(+)

diff --git a/drivers/platform/x86/intel/int3472/discrete.c b/drivers/platform/x86/intel/int3472/discrete.c
index 9c67a8fd6642..c5dfaa912af1 100644
--- a/drivers/platform/x86/intel/int3472/discrete.c
+++ b/drivers/platform/x86/intel/int3472/discrete.c
@@ -142,6 +142,33 @@ struct int3472_gpio_map {
 };
 
 static const struct int3472_gpio_map int3472_gpio_map[] = {
+	{	/* Galaxy Book 12: autofocus actuator power */
+		.hid = "SONY258A",
+		.type_from = 0x0e,
+		.type_to = INT3472_GPIO_TYPE_HOTPLUG_DETECT,
+		.con_id = "af-enable",
+	},
+	{	/* Galaxy Book 12: legacy Intel "IO" rail feeds IMX258 VIF */
+		.hid = "SONY258A",
+		.type_from = 0x0f,
+		.type_to = INT3472_GPIO_TYPE_POWER_ENABLE,
+		.con_id = "vif",
+		.enable_time_us = GPIO_REGULATOR_ENABLE_TIME,
+	},
+	{	/* Galaxy Book 12: legacy Intel "Avdd" rail */
+		.hid = "SONY258A",
+		.type_from = INT3472_GPIO_TYPE_DOVDD,
+		.type_to = INT3472_GPIO_TYPE_POWER_ENABLE,
+		.con_id = "vana",
+		.enable_time_us = GPIO_REGULATOR_ENABLE_TIME,
+	},
+	{	/* Galaxy Book 12: legacy Intel "Core" rail */
+		.hid = "SONY258A",
+		.type_from = 0x11,
+		.type_to = INT3472_GPIO_TYPE_POWER_ENABLE,
+		.con_id = "vdig",
+		.enable_time_us = GPIO_REGULATOR_ENABLE_TIME,
+	},
 	{	/* mt9m114 designs declare a powerdown pin which controls the regulators */
 		.hid = "INT33F0",
 		.type_from = INT3472_GPIO_TYPE_POWERDOWN,
-- 
2.50.1
